src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt

```diff
@@ -0,0 +1,115 @@
+import org.rust.lang.core.psi.ext.childOfType
+import org.rust.lang.core.psi.ext.endOffset
+import org.rust.lang.core.psi.ext.parentDotExpr
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.type
+import javax.swing.JComponent
+
+class RsChainMethodTypeHintsProvider : InlayHintsProvider<RsChainMethodTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = KEY
+
+    override val name: String get() = "Chain method hints"
+
+    override val previewText: String
+        get() = """
+            let foo: Result<Option<i32>, &str> = Ok(Some(10i32));
+            foo
+              .map_err(|_| 0_u32)
+              .unwrap()
+              .and_then(|_| Some("foo"))
+              .unwrap();
+            """.trimIndent()
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        override fun createComponent(listener: ChangeListener): JComponent {
+            return panel {}
+        }
+    }
+
+    override fun createSettings(): Settings = Settings()
+
+    override fun getCollectorFor(file: PsiFile, editor: Editor, settings: Settings, sink: InlayHintsSink): InlayHintsCollector? =
+        object : FactoryInlayHintsCollector(editor) {
+
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, true)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (DumbService.getInstance(file.project).isDumb) return true
+                if (element !is RsElement) return true
+                if (!element.isEnabledByCfg) return true
+                if (element !is RsMethodCall) return true
+                if (!element.isLastInChain) return true
+
+                val chain = collectChain(element)
+                var lastType: Ty? = null
+                for (call in chain.dropLast(1)) {
+                    if (call.type != lastType) {
+                        presentTypeForMethodCall(call)
+                        lastType = call.type
+                    }
+                }
+
+                return true
+            }
+
+            private fun presentTypeForMethodCall(call: RsMethodCall) {
+                val project = call.project
+                val type = call.type
+                val presentation = typeHintsFactory.typeHint(type)
+                val finalPresentation = presentation.withDisableAction(project)
+                sink.addInlineElement(call.endOffset, true, finalPresentation)
+            }
+        }
+
+    private fun InlayPresentation.withDisableAction(project: Project): InsetPresentation = InsetPresentation(
+        MenuOnClickPresentation(this, project) {
+            listOf(InlayProviderDisablingAction(name, RsLanguage, project, key))
+        }, left = 1
+    )
+
+    class Settings
+
+    companion object {
+        private val KEY: SettingsKey<Settings> = SettingsKey("rust.chain-method.hints")
```

Comment:
`rust` prefix looks redundant - `SettingsKey` add it itself. See SettingsKey#getFullId`

Reply:
`RsInlayTypeHintsProvider` uses `"rust.type.hints"`, that's why I have included it. Should I remove the prefix?

Reply:
Yep, it was a mistake but if you remove prefix in `RsInlayTypeHintsProvider ` it will affect all users who changed the corresponding settings so I think we shouldn't do it